pkg/types/chain_reader.go

```diff
@@ -51,21 +83,260 @@ type BoundContract struct {
 	Pending bool
 }
 
-type Event struct {
-	ChainID           string
-	EventIndexInBlock string
-	Address           string
-	TxHash            string
-	BlockHash         string
-	BlockNumber       int64
-	BlockTimestamp    time.Time
-	CreatedAt         time.Time
-	Keys              []string
-	Data              []byte
+// Head TODO this is a general chain agnostic Head, its copied from mercury Chain Reader.
+// Mercury Chain Reader should be merged with this Chain Reader, so we won't have duplicate Head definitions.
+type Head struct {
+	Number    uint64
+	Hash      []byte
+	Timestamp uint64
+}
+
+type Sequence struct {
+	// TODO SequenceCursor, this should be a unique sequence identifier that chain reader impl. understands.
+	// This way we can retrieve past/future sequences (EVM log events) very granularly, but still hide the chain detail.
+	SequenceCursor string
+	Head
+	Data any
+}
+
+type ComparisonOperator int
+
+const (
+	Eq ComparisonOperator = iota
+	Neq
+	Gt
+	Lt
+	Gte
+	Lte
+)
+
+type SortDirection int
+
+const (
+	Asc SortDirection = iota
+	Desc
+)
+
+type SortBy interface {
+	GetDirection() SortDirection
+}
+
+type LimitAndSort struct {
+	SortBy []SortBy
+	Limit  uint64
+}
```

Comment:
I have a feeling we may want to expand this later to allow for more result modification options.  Maybe we should name it something more general like `OutputMods`, `ResultMods` or `ResultTransforms`, `SequenceTransforms`, `OutputOpts`, `ResultFormatOpts`, etc?

It could be anything that rearranges the results or filters them in some way.  For example, maybe you want to drop both the first 5 results and the last 5 results, like `Head` & `Tail` operations instead of just `Limit`.  Or what if you want only results that when sorted are at an odd or even index? Can't think of specific use cases for any of these off the top of my head, just seems like at some point we're bound to want to add something beyond `Sort` & `Limit`

Reply:
I can't think of any useful use case other than `LimitAndSort`, so thats why I called it like that.

Reply:
I happen to agree with @ilija42 here.  It's difficult in SQL to drop the first five results (could do it with joins and some hackery).  I can't think of when you would want to drop head and tail during a query.  Dropping head usually is just continuing from a cursor, dropping tail is just a lower limit.

Reply:
> It's difficult in SQL to drop the first five results (could do it with joins and some hackery).

I thought we were trying to design this in a way that doesn't assume the cache will be a sql db... it could be in-memory storage, key-value storage, etc. right? For now, since we are using postgres on the backend it's a practical concern... but we could still implement it in the relay if there was a need, by dropping some results after the db query comes back, before it gets sent back over gRPC.  (I guess the product plugin itself could also filter the results after receiving them though, which wouldn't be too much different.)

> Dropping head usually is just continuing from a cursor, dropping tail is just a lower limit.

Ah, continuing from a cursor actually seems like it might be an important use though, isn't it?  If there are a large number of events, maybe the product wants to fetch them in batches of 1000 at a time or something. How would you say "give me the next 1000 events that match this filter"? If we implement it with cursors, then we'd have to introduce some sort of statefulness where you make one query and it sends you back a token (like the "fingerprints" in the Tron API) which can be used to request the next one.  If we implement it in terms of Head & Tail (allowing you to say something like "give me events 5000 - 6000 when sorted by this key" with the `SortAndLimit` param), that would allow us to keep the API stateless... which seems better, right?

Reply:
In postgresql, it's not terribly hard to have an upper and lower limit like that... you can use `ROW_NUMBER() OVER(ORDER BY key) as row_id` to generate a column in any query that numbers the row results, then use that as a subquery of a query with a  WHERE clause which filters by `row_id >= min AND row_id <= max`

Reply:
Admittedly, even if we add support for that, the name `SortAndLimit` isn't too bad... you're still sorting and limiting the results. I'm oikay with keeping it.

Reply:
It doesn't assume what the underlying db impl. is, but if we know that it would mostly be SQL then it is important that something is hard to do with SQL.

I was imagining the cursor to be a mandatory unique identifier attached  to every sequence. This would be just a concat of all things that make it unique (blocknumber-hash-topicId-whatever). This cursor could be added as a `CursorFilter` to `QueryFilter` or to the `LimitAndSort`.  Functionally it would do the same thing where we can ask to continue from the cursor or limit to next/previous 1000 sequences.